### ISSUE TITLE
Interesting issue with references

### ISSUE THREAD

**Issue by jnbrq** [ThePhD/sol2#495]
Consider the following code:

```
#include <string>
#include <sol.hpp>

void g2(std::string &) {}
// this fails
// but if I pass a const std::string &, it is fine

int main() {

    sol::state lua;
    lua["g2"] = &g2;
    return 0;
}
```

It fails. Actually, I was having problems with a specialized string class wxString (from wxWidgets library), and I tried to isolate the problem. This is what I got.

I am using the latest version of the develop branch.

Error:
```
In file included from /home/janberq/Some Important Stuff/Sol2 Bug-alike?/main.cpp:2:0:
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp: In instantiation of ‘static decltype(auto) sol::wrapper<F, typename std::enable_if<std::is_function<typename std::remove_pointer<typename std::remove_cv<typename std::remove_reference<_From>::type>::type>::type>::value, void>::type>::call(F&, Args&& ...) [with Args = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}; F = void (*)(std::__cxx11::basic_string<char>&)]’:
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:9202:16:   required from ‘decltype(auto) sol::wrapper<F, typename std::enable_if<std::is_function<typename std::remove_pointer<typename std::remove_cv<typename std::remove_reference<_From>::type>::type>::type>::value, void>::type>::caller::operator()(F&, Args&& ...) const [with Args = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}; F = void (*)(std::__cxx11::basic_string<char>&)]’
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:8895:33:   required from ‘static decltype(auto) sol::stack::stack_detail::evaluator::eval(sol::types<>, std::index_sequence<>, lua_State*, int, sol::stack::record&, Fx&&, Args&& ...) [with Fx = sol::wrapper<void (*)(std::__cxx11::basic_string<char>&), void>::caller; Args = {void (*&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}; std::index_sequence<> = std::integer_sequence<long unsigned int>; lua_State = lua_State]’
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:8900:17:   required from ‘static decltype(auto) sol::stack::stack_detail::evaluator::eval(sol::types<Arg, Args ...>, std::index_sequence<I, Is ...>, lua_State*, int, sol::stack::record&, Fx&&, FxArgs&& ...) [with Fx = sol::wrapper<void (*)(std::__cxx11::basic_string<char>&), void>::caller; Arg = std::__cxx11::basic_string<char>&; Args = {}; long unsigned int I = 0ul; long unsigned int ...Is = {}; FxArgs = {void (*&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)}; std::index_sequence<I, Is ...> = std::integer_sequence<long unsigned int, 0ul>; lua_State = lua_State]’
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:8921:5:   required from ‘void sol::stack::stack_detail::call(sol::types<void>, sol::types<Args ...>, std::index_sequence<I ...>, lua_State*, int, Fx&&, FxArgs&& ...) [with bool checkargs = true; long unsigned int ...I = {0ul}; Args = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&}; Fx = sol::wrapper<void (*)(std::__cxx11::basic_string<char>&), void>::caller; FxArgs = {void (*&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)}; std::index_sequence<I ...> = std::integer_sequence<long unsigned int, 0ul>; lua_State = lua_State]’
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:8945:34:   required from ‘void sol::stack::call(sol::types<void>, sol::types<Args ...>, lua_State*, int, Fx&&, FxArgs&& ...) [with bool check_args = true; Args = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&}; Fx = sol::wrapper<void (*)(std::__cxx11::basic_string<char>&), void>::caller; FxArgs = {void (*&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)}; lua_State = lua_State]’
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:8967:20:   [ skipping 20 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:16125:13:   required from ‘sol::basic_table_core<<anonymous>, T>& sol::basic_table_core<<anonymous>, T>::set(Args&& ...) [with Args = {std::tuple<const char (&)[3]>&, sol::function_arguments<sol::function_sig<>, void (*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)>}; bool top_level = true; base_type = sol::reference]’
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:16279:7:   required from ‘void sol::basic_table_core<<anonymous>, T>::set_fx(sol::types<>, Key&&, Fx&&, Args&& ...) [with Fx = void (*)(std::__cxx11::basic_string<char>&); Key = std::tuple<const char (&)[3]>&; Args = {}; typename std::enable_if<std::integral_constant<bool, (! sol::meta::all<sol::meta::is_specialization_of<sol::overload_set, typename std::remove_cv<typename std::remove_reference<Arg>::type>::type> >::value)>::value, sol::meta::enable_t>::type <anonymous> = (sol::meta::type)0; bool top_level = true; base_type = sol::reference]’
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:16253:10:   required from ‘sol::basic_table_core<<anonymous>, T>& sol::basic_table_core<<anonymous>, T>::set_function(Key&&, Args&& ...) [with Key = std::tuple<const char (&)[3]>&; Args = {void (*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)}; bool top_level = true; base_type = sol::reference]’
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:11918:4:   required from ‘sol::proxy<Table, Key>& sol::proxy<Table, Key>::set_function(Args&& ...) [with Args = {void (*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)}; Table = sol::basic_table_core<true, sol::reference>&; Key = const char (&)[3]]’
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:11924:23:   required from ‘sol::proxy<Table, Key>& sol::proxy<Table, Key>::operator=(U&&) [with U = void (*)(std::__cxx11::basic_string<char>&); typename std::enable_if<sol::meta::all<std::integral_constant<bool, (! sol::is_lua_reference<typename sol::meta::unwrap_unqualified<Fx>::type>::value)>, sol::meta::is_callable<typename sol::meta::unwrap_unqualified<Fx>::type> >::value, sol::meta::enable_t>::type <anonymous> = (sol::meta::type)0; Table = sol::basic_table_core<true, sol::reference>&; Key = const char (&)[3]]’
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/main.cpp:11:15:   required from here
/home/janberq/Some Important Stuff/Sol2 Bug-alike?/external/sol2/single/sol/sol.hpp:9196:13: error: invalid initialization of non-const reference of type ‘std::__cxx11::basic_string<char>&’ from an rvalue of type ‘std::__cxx11::basic_string<char>’
    return fx(std::forward<Args>(args)...);
```

My system specs:
```
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/5/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.4.0-6ubuntu1~16.04.4' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.4)
```

**Comment by ThePhD** [comment]
Strings are primitives in Lua. You can't take a reference to a primitive type, because they exhibit copy semantics.

This compiles in VC++ because of a (bad?) Microsoft-specific extension. G++ errors it because it doesn't allow the value type of `std::string` we create (that is an unnamed temporary passed to your function) to be captured with `std::string&`, because it would erroneously give you the idea that you're modifying the memory of a string directly (you're not, it's a fresh temporary).

Take a `std::string const&` or a `std::string`. If you want to avoid a copy, using `sol::string_view`. If you know it's a c-style string and has no `\0`'s in the body, you are permitted to use `const char*`. Note that all primitives from Lua (all integral types and strings) more or less exhibit value semantics.

As a side note, you should write a specialized getter and setter for your `wx` string class. In fact, I see you already have previously in issue #140 -- has something changed?

**Comment by jnbrq** [comment]
Ohh. Yeah, you're right. That was a silly mistake :D Whatever, my real problem was, my code working with a previous version of sol2 is now broken. I was trying to investigate that. I will investigate it further and close this issue and open a new one, or continue from here.

**Comment by jnbrq** [comment]
Well, I found the problem, I was looking for it in the wrong place. I am trying to register a class with more than 50 functions. The problem is the recursion depth of the C++ compiler doesn't allow doing so. But my code used to work before, most probably changes in the library increased the required recursion depth. So, I increased it and no problem now! Maybe, it might be good to put a note regarding recursion depth. Thank you :).
